extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+# workspace/reference extension to LSP
+
+The `workspace/reference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references that code inside of the workspace root makes to code symbols that are external to the workspace root (e.g. dependencies).
+
+Use case: clients of a language server can invoke `workspace/reference` in order to find external references to e.g. imported dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.URI` which is located within the workspace.
+- Excluding any `ReferenceInformation.Name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.Location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace at which the `symbol` has been referenced.
+	 */
+	location: Location;
+
+	/**
+	 * The name of the symbol that is being referenced. For example:
+	 *
+	 * 	(Go) "ServeHTTP"
+	 *  (JS) "render"
+	 *
+	 */
+	name?: string;
+
+	/**
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
```

Comment:
> I think we do need that extra metadata key-value map that I proposed (both here and in workspace/symbol)

I think I still don't have a clear picture of this proposal, but I would like to. I remember you talking about it with me on Slack, but in my mind it is still some form of "let's keep magical metadata around and it will disambiguate things". Could you help elaborate more on what it is and how exactly it would function?

> With kind or the arbitrary key-value metadata, we could reduce ambiguous resolutions to a small enough number, I believe.

The point is that for "Top Defs" on the repo landing page (https://sourcegraph.com/github.com/golang/go/-/info), at least with the way things currently exist, we cannot have _any_ ambiguousness. We have to have some level of uniqueness, because the DB stores one thing and we must somehow map that back to a specific location. The current approach is `(uri, name, containerName)` where `uri` must include either a dir or file. We can offer disambiguation for def landing, surely, but for repo landing we've got ~20 definitions to resolve -- asking the user to disambiguate all of those would be madness :)